### DISCUSSION TITLE
Have I got all the entities?

### DISCUSSION THREAD

**Discussion by Sboshoff76** [celsworth/lxp-bridge#177] [General]
Hi

First off thanks very much for creating this. I am new to HA and have been finding my way around things (very slowly) :) so excuse my ignorance.

I Installed V0.10 and the below is what was exposed to me. 
![image](https://github.com/celsworth/lxp-bridge/assets/68126343/cd1edae3-0f63-429c-ba7c-24644eb60b0e)
![image](https://github.com/celsworth/lxp-bridge/assets/68126343/969863a4-2b84-4f3b-b37a-e085d3720182)
![image](https://github.com/celsworth/lxp-bridge/assets/68126343/e2f5d2b8-4227-4592-9d1c-ffd0f5a5d4e1)

I then tinkered with MQTTExplorer and found further values that were useful for me in the /inputs/all register which I started using by creating some manual sensors (like the the instantaneous power draws to/from grid or battery or consumption).

Did I do something wrong that all these entities were not exposed initially?

And then secondly, I am yet to find where to read my battery stats... any ideas?

Thanks again!

**Comment by celsworth** [comment]
Hi :)

"Power from Grid" and "Power to Grid" should certainly be there - they're exposed to HA here - https://github.com/celsworth/lxp-bridge/blob/master/src/home_assistant.rs#L106-L107

Indeed I can see Power to Grid in your screenshot, but not Power from Grid, is that one missing? Battery Discharge should be there too, but I don't see that in your screenshot. Be odd if just a couple were missing.

So the way this works is lxp-bridge publishes messages on startup that HomeAssistant then parses as "there's a sensor available and here's a description of where to find it". The one for "Battery Discharge" lives at `homeassistant/sensor/lxp_XXXXXXXXXX/p_discharge/config` where XXXXXXXXXXX are replaced with your inverter serial number. These are published retained so you should be able to do something like:

```
mosquitto_sub -v -t 'homeassistant/sensor/#'
```

To get a complete list of what we've published for HA to see. Might be a worth look through those for starters. If the `p_discharge` one is there then HA is ignoring it for some reason, maybe check logs of HA to see if there's any clues.

For battery stats can you elaborate what exactly you're wanting?

**Comment by Sboshoff76** [comment]
Appreciate the swift repsonse.

Yeah I'm not sure why some of mine are missing. I've imported most of the "missing" ones with code like this:

    - name: "11-40 Power into Battery"
      state_topic: lxp/BAXXXXXXXXX/inputs/all
      state_class: measurement
      device_class: power
      value_template: "{{ value_json.p_charge }}"
      unit_of_measurement: "W"
      unique_id: lxp_BAXXXXXXXX_p_charge

My complete "DIY" list is as follows:
p_to_user
p_charge
p_discharge
p_inv
p_eps (yours is measuring VA power whilst all the others are in W, so I pulled this in myself again)
p_rec
e_to_user_day
e_eps_day
e_inv_day
ep_pv_day_1
ep_pv_day_2
e_chg_day
e_dischg_day

For battery stats you bring in SOC, but other things like cycle count or battery health would be good. You already bring in an overall battery voltage reading but for people like me that have multiple batteries in parallel the individual battery voltages would be great.

**Comment by celsworth** [comment]
Something a little odd going on there as some of the ones you mention you had to setup are definitely already in (see list from previous comment). For example `p_eps` is already there which is W (active power), but there's also `s_eps` which is apparent power (VA). But again I don't see the active power in your list, only apparent.. Not sure what the problem is, to be honest :(

For adding more battery stats, we're a little limited on what HA supports here I think. I think we can only send the things that have a type listed at https://developers.home-assistant.io/docs/core/entity/sensor/#available-device-classes

So for example, SOC is `SensorDeviceClass.BATTERY`, but there is nothing in that list that I can see to allow us to send a SOH and have HA make sense of it..similar story for cycles. I'll keep digging though and see if I can work anything out.

**Reply by Sboshoff76** [reply]
Maybe something went weird with my install. I'll load up a HA VM and do a fresh install to see if something different is exposed.

**Comment by Sboshoff76** [comment]
FYI. Diagnosing with this app it seems only 30 topics got pulled in. It seems to correlate with what I have/don't have.
PS. I have 2 units running in parallel

![image](https://github.com/celsworth/lxp-bridge/assets/68126343/58ee6dc2-97fe-430a-9de0-f2247390ea1d)

**Reply by celsworth** [reply]
Ignore previous comments about queue size, don't think it's relevant.

So in this list I do see `p_to_user` and `p_to_grid` and `p_eps`, to pick 3.. those are the ones you don't have in your HA, but they are present in MQTT?

**Comment by celsworth** [comment]
So next line of thought, if you enable debug logging in lxp-bridge and restart, it should tell you which messages are being published on startup, like this:

```
[2023-07-13T21:29:48.796 DEBUG rumqttc::state] Publish. Topic = homeassistant/sensor/lxp_XXXXXXXXXX/p_eps/config, Pkid = 22, Payload Size = 354
```

Do you see those for the missing sensors?

**Comment by Sboshoff76** [comment]
I do not have:
self.power("p_to_user", "Power from Grid")?,
self.power("p_eps", "Active EPS Power")?,

I only have:
self.power("p_to_grid", "Power to Grid")?,


Where do I set the log level? Is it the RUST_LOG setting in the service file? If so I can't seem to find that (or any files pertaining to lxp-bridge). Is it because I'm running the OS version in Hyper-V ?

**Comment by Sboshoff76** [comment]
I deployed a test instance , linking it to only 1 of my 2 inverters and the missing ones came through. I did notice that there are still some entities missing that I require though and they might be beneficial for other users, so going to list them here and hopefully you can pull them in.

e_chg_day
e_dischg_day
e_eps_all
e_eps_day
e_inv_all
e_inv_day
e_pv_day
e_pv_day_1
e_pv_day_2
e_pv_day_3
e_rec_all
e_rec_day
e_to_grid_day
e_to_user_day

p_inv
p_rec

Thanks!

**Comment by Sboshoff76** [comment]
Thanks for merging my pull request. Seeing as there is no release yet, how can I switch over to the dev branch to give them a spin?

**Reply by celsworth** [reply]
If you're running lxp-bridge in a standalone Docker container, you can switch from the default `latest` tag to the `dev` tag which contains the latest build every time master gets an update: https://hub.docker.com/layers/celsworth/lxp-bridge/dev/images/sha256-de8251346d4363502318d874623485995f69e644bb129ab983b231d695e8342f?context=repo

If you're running it via the HA addon I don't think there's a way at the moment. The `addon/config.yaml` file directs HA to install the v0.11.0 Docker tag.. not sure if I can make multiple versions of this or how you would choose one, HA addons aren't my strong point I'm afraid. Would be nice to improve this so I wonder if we could just copy the `addon` directory to an `addon-dev` with appropriate changes - just need to work out how to get the HA button to use it..

**Reply by Sboshoff76** [reply]
Ah k, that's what I thought. Yes I run the addon so have no access to the underlying files.

**Reply by Sboshoff76** [reply]
What if you create a whole new addon and just call it lxpbridge-dev like esphome does? Or does that mess too much with the backend and configs on your end?

**Comment by Sboshoff76** [comment]
@celsworth I'm using your integration for displaying some power flows in a 3rd party lovelace card. Some of the entities like battery and grid needs a nett figure where a negative means discharge/export and a positive is charge/import. It's easy to do with a template sensor but do you think it's something you can include  in your integration as then users wouldn't need to manually do it to use the card. I'll be happy to provide my templates if you want.

**Reply by celsworth** [reply]
Yeah that's pretty easy, I can add that in today. So I'm thinking to add:

* `p_grid` - the result of `p_to_user - p_to_grid`
* `p_battery` - the result of `p_charge - p_discharge`

So if `p_charge` is 0 and `p_discharge` is 300 then the new `p_battery` would be -300. Is that right?

**Reply by Sboshoff76** [reply]
Spot on!

**Reply by celsworth** [reply]
Added in #183!

**Comment by Sboshoff76** [comment]
@celsworth While you're making changes can I be cheeky and ask for something else?

My inverter is in an outbuilding and I'd like to duplicate the display from the inverter onto a tablet in the house. I have most of the data & flows already except for the following:

Error/Warning Codes (Number 4 below)
% of power (Number 9 below)
![image](https://github.com/celsworth/lxp-bridge/assets/68126343/852e8097-1981-468d-986a-734c2e7be1fa)

Think you can grab those for us?

**Reply by celsworth** [reply]
Errors and warnings should be fine, we know the registers for those.

% of power might be trickier, I've not seen a register for that so presumably we need to do a calculation of current power divided by max power. But that's a problem because I'm not seeing a register that exposes the maximum power of the inverter..

**Reply by Sboshoff76** [reply]
Yeah I figured the last one might be a bother. Could you lookup a max value based on the model of inverter identified and then do the calc off that?

**Reply by pjv** [reply]
@Sboshoff76 Do you have any HA dashboard configs / cards / templates / etc that you could share for your tablet replication of the inverter LCD?

**Comment by snowjog** [comment]
@celsworth I seem to be struggling retrieving the current inverter power (p_inv) I can see it in the MQTT messages, and I found reference to it in your home_assistant.rs file with an entity name of "Inverter Power" but that attribute doesn't seem to be exposed/exist when I go to add it to a dashboard card, or whatever - am I doing something wrong?

![ha_lxp_bridge](https://github.com/celsworth/lxp-bridge/assets/9728391/d295238d-353e-4a7f-adc3-12ac7e95e40b)

Thanks

**Reply by celsworth** [reply]
Hi!

If you're using the release version of lxp-bridge, this will be why - `p_inv` was only added very recently in #181 which hasn't made it into a release yet.

It's probably about time for a new release so I'll do this soon.

**Reply by snowjog** [reply]
Hi Chris thanks for the latest release - that does now show p_inv. I've also been trying in vain to set my discharge cutoff % to the same as the battery soc% using an automation. The setting of an entity doesn't seem to like template stuff in there {{ states('example') }} so I tried calling a service (MQTT) and setting the following: 

lxp/baXXXXXXXX/hold/105 QOS 0, retain, with a value of 20. This now always appears when I do an MQTT listen. 

How do I stop that? Also I have in lxp/BAXXXXXXXX/inputs/all the following: "internal_fault": 12544, not sure if that was there before, or if it's something I've done.

TIA
Jez

**Reply by celsworth** [reply]
Hello :)

I'm afraid I can't help much with the HA automations, I'm not a HA user primarily. But hopefully someone else can chime in with that.

So there's a few points to address here - firstly the retain message stuff. When sending a message to lxp-bridge to set a register, you don't want to set the retain bit. That will mean the message is stored on the broker and sent to lxp-bridge every time it restarts, which will reset the register to the value contained in the retained message. This probably isn't what you want.

To clear a retained message from a broker, you can usually just send an empty payload to the topic. So using `mosquitto_pub`, the Linux commandline tool:

```
mosquitto_pub -t lxp/baXXXXXXXX/hold/105 -m ''
```

However, the topic you've used is not right to set a register. That's the one that lxp-bridge publishes to when the value of the register is received from the inverter. It is marked retain so anyone querying the broker can get the last known value of that register without having to wait for an update for it. Since most of them change very infrequently, it is *probably* still the current value, especially if lxp-bridge is left running.

To set a new value, you want to use `lxp/cmd/baXXXXXXXX/set/hold/105` (see https://github.com/celsworth/lxp-bridge/wiki/MQTT#lxpcmddatalogsetholdx-payload--int)

Finally the `internal_fault`. I wouldn't worry too much about this, you are right it wasn't there before, I've just exposed it in the most recent version, but unfortunately what any of the values mean isn't known. There's no documentation about it :( So if your inverter appears to be working normally then I really wouldn't worry about there being a non-zero value in  there.

**Comment by Sboshoff76** [comment]
Hiya @celsworth me again with another request! Any chance you can reveal the below 2 input registers?

81 | max_chg_curr | 0.01A | BMS max charge current
82 | max_dischg_curr | 0.01A | BMS max discharge current

When the batteries are almost full it throttles the charging current meaning I'd like to make use of automations to turn on other items and use PV that is available.

Thanks

**Reply by celsworth** [reply]
Yep, think these should be fairly easy, I'll add them today.

**Reply by celsworth** [reply]
Done - `dev` Docker tag should be updated soon if you'd like to give it a try (you can install the dev image from the HA repo now, give it a refresh).

**Reply by Sboshoff76** [reply]
Awesome.. Will give it a spin soon. Thanks!

PS. You've converted me to dev and docker ;)

**Reply by Sboshoff76** [reply]
![image](https://github.com/celsworth/lxp-bridge/assets/68126343/1560e2d5-bd2f-49b0-a40f-22d2a7aace10)

Thanks Chris!
